runtime/signals_osdep.h

```diff
@@ -186,15 +186,16 @@
 #elif defined(TARGET_i386) && defined(SYS_linux_elf)
 
   #define DECLARE_SIGNAL_HANDLER(name) \
-    static void name(int sig, struct sigcontext context)
+    static void name(int sig, siginfo_t * info, ucontext_t * context)
 
   #define SET_SIGACT(sigact,name) \
-     sigact.sa_handler = (void (*)(int)) (name); \
-     sigact.sa_flags = 0
+     sigact.sa_sigaction = (void (*)(int,siginfo_t *,void *)) (name); \
+     sigact.sa_flags = SA_SIGINFO
 
-  #define CONTEXT_FAULTING_ADDRESS ((char *) context.cr2)
-  #define CONTEXT_PC (context.eip)
-  #define CONTEXT_SP (context.esp)
+  typedef greg_t context_reg;
+  #define CONTEXT_PC (context->uc_mcontext.gregs[REG_EIP])
+  #define CONTEXT_SP (context->uc_mcontext.gregs[REG_ESP])
+  #define CONTEXT_FAULTING_ADDRESS ((char *)context->uc_mcontext.cr2)
```

Comment:
Why `uc_mcontext.cr2` rather than `info->si_addr`?

Reply:
Conservatism.  Using CR2 is the way it was done before for i386 and the way it is still done for AMD64.  If it ain't broke...

Reply:
>  If it ain't broke...

Then fix it until it is?

(Eventually, it would be nice to use the POSIX `info->si_addr` everywhere rather than per-platform tricks, but this may not be the PR for that change)